aw_client/client.py

```diff
@@ -72,72 +60,155 @@ def _get(self, endpoint: str) -> Optional[req.Response]:
+                return True
+
+    #
+    #   Failed request queue handling
+    #
+
+    def _queue_failed_request(self, endpoint: str, data: dict):
+        # Find failed queue file
+        entry = {"endpoint": endpoint, "data": data}
+        self.queue_file_lock.acquire()
+        with open(self.queue_file, "a+") as queue_fp:
+            queue_fp.write(json.dumps(entry) + "\n")
+        self.queue_file_lock.release()
+
+    def _post_failed_requests(self):
+        failed_requests = []
+        self.queue_file_lock.acquire()
+        with open(self.queue_file, "r") as queue_fp:
+            for request in queue_fp:
+                failed_requests.append(json.loads(request))
+            if len(failed_requests) != 0:
+                open(self.queue_file, "w").close()  # Clear file
+                logger.info("Sending {} failed events: {}".format(len(failed_requests), failed_requests))
+                for request in failed_requests:
+                    self._post(request['endpoint'], request['data'])
+        self.queue_file_lock.release()
+
+    #
+    #   Connection methods
+    #
+
+    def connect(self):
+        if self._try_connect():
+            self.connected = True
         else:
-            # Create bucket
-            endpoint = "buckets/{}".format(bucket_id)
-            data = {
-                'client': self.client_name,
-                'hostname': self.client_hostname,
-                'type': event_type,
-            }
-            try:
-                self._post(endpoint, data)
-            except req.RequestException as e:
-                self.logger.error("Unable to create bucket: {}".format(e))
+            logger.warning("Can't connect to aw-server, will queue events until connection is available")
+            if self.reconnect_thread == None:
+                self.reconnect_thread = ReconnectThread(self).start()
+
+    def _try_connect(self):
+        try:
+            self._create_buckets()
+            self._post_failed_requests()
             return True
+        except req.RequestException as e:
+            return False
+
+    def _connection_lost(self):
+        """
+            Call this when we lose connection to the server
+        """
+        self.connected = False
+        if self.reconnect_thread == None:
+            logger.warning("Connection to aw-server lost, will queue events until connection is available again")
+            self.reconnect_thread = ReconnectThread(self).start()
 
 
-class QueueTimerThread(threading.Thread):
+class ReconnectThread(threading.Thread):
     def __init__(self, client):
         threading.Thread.__init__(self)
         self.daemon = True
         self.client = client
 
     def run(self):
-        while True:
-            time.sleep(180)
-            self.client._post_failed_events()
+        while not self.client.connected:
+            if self.client._try_connect():
+                self.client.connected = True
```

Comment:
Changing the state of self.client from here is an antipattern, you should probably do it in `self.client._try_connect()` instead.